batch.go

```diff
@@ -0,0 +1,273 @@
+		URL:      BatchEndpointChatCompletions,
+	})
+}
+
+func (r *UploadBatchFileRequest) AddCompletion(customerID string, body CompletionRequest) {
+	r.Lines = append(r.Lines, BatchCompletionRequest{
+		CustomID: customerID,
+		Body:     body,
+		Method:   "POST",
+		URL:      BatchEndpointCompletions,
+	})
+}
+
+func (r *UploadBatchFileRequest) AddEmbedding(customerID string, body EmbeddingRequest) {
+	r.Lines = append(r.Lines, BatchEmbeddingRequest{
+		CustomID: customerID,
+		Body:     body,
+		Method:   "POST",
+		URL:      BatchEndpointEmbeddings,
+	})
+}
+
+// UploadBatchFile — upload batch file.
+func (c *Client) UploadBatchFile(ctx context.Context, request UploadBatchFileRequest) (File, error) {
+	if request.FileName == "" {
+		request.FileName = "@batchinput.jsonl"
+	}
+	return c.CreateFileBytes(ctx, FileBytesRequest{
+		Name:    request.FileName,
+		Bytes:   request.MarshalJSONL(),
+		Purpose: PurposeBatch,
+	})
+}
+
+type CreateBatchWithUploadFileRequest struct {
+	Endpoint         BatchEndpoint  `json:"endpoint"`
+	CompletionWindow string         `json:"completion_window"`
+	Metadata         map[string]any `json:"metadata"`
+	UploadBatchFileRequest
+}
+
+// CreateBatchWithUploadFile — API call to Create batch with upload file.
+func (c *Client) CreateBatchWithUploadFile(
+	ctx context.Context,
+	request CreateBatchWithUploadFileRequest,
+) (response BatchResponse, err error) {
+	var file File
+	file, err = c.UploadBatchFile(ctx, UploadBatchFileRequest{
+		FileName: request.FileName,
+		Lines:    request.Lines,
+	})
+	if err != nil {
+		err = errors.Join(ErrUploadBatchFileFailed, err)
+		return
+	}
+	return c.CreateBatch(ctx, CreateBatchRequest{
+		InputFileID:      file.ID,
+		Endpoint:         request.Endpoint,
+		CompletionWindow: request.CompletionWindow,
+		Metadata:         request.Metadata,
+	})
+}
+
+// RetrieveBatch — API call to Retrieve batch.
+func (c *Client) RetrieveBatch(
+	ctx context.Context,
+	batchID string,
+) (response BatchResponse, err error) {
+	urlSuffix := fmt.Sprintf("%s/%s", batchesSuffix, batchID)
+	req, err := c.newRequest(ctx, http.MethodGet, c.fullURL(urlSuffix))
+	if err != nil {
+		return
+	}
+	err = c.sendRequest(req, &response)
+	return
+}
+
+// CancelBatch — API call to Cancel batch.
+func (c *Client) CancelBatch(
+	ctx context.Context,
+	batchID string,
+) (response BatchResponse, err error) {
+	urlSuffix := fmt.Sprintf("%s/%s/cancel", batchesSuffix, batchID)
+	req, err := c.newRequest(ctx, http.MethodPost, c.fullURL(urlSuffix))
+	if err != nil {
+		return
+	}
+	err = c.sendRequest(req, &response)
+	return
+}
+
+type ListBatchResponse struct {
+	httpHeader
+	Object  string  `json:"object"`
+	Data    []Batch `json:"data"`
+	FirstID string  `json:"first_id"`
+	LastID  string  `json:"last_id"`
+	HasMore bool    `json:"has_more"`
+}
+
+// ListBatch API call to List batch.
+func (c *Client) ListBatch(ctx context.Context, after *string, limit *int) (response ListBatchResponse, err error) {
```

Comment:
Should we use the [Pagination](https://github.com/sashabaranov/go-openai/blob/211cb49fc22766f4174fef15301c4d39aef609d3/run.go#L210) struct here?

Reply:
The request parameters provided by openai are inconsistent with the Pagination structure
https://platform.openai.com/docs/api-reference/batch/list